# Exercícios - Dados Bolsa Família (Métricas com Pandas)

Lista de exercícios com foco em contagens, valores distintos, média, mediana, mínimo e máximo.


## DADOS

Os dados para essa atividade encontram-se em:

https://drive.google.com/file/d/13KRdiYnQikDx3pxPKGUTeejOz1fLHjzm/view?usp=sharing

In [16]:
import pandas as pd
import gdown
import tempfile
import os
import uuid

#Baixando o dataset do Google Drive e lendo-o em um DataFrame
temp_dir = tempfile.gettempdir()
temp_path = os.path.join(temp_dir, f"dataset_{uuid.uuid4().hex}.parquet")
try:
    gdown.download(id='13KRdiYnQikDx3pxPKGUTeejOz1fLHjzm', output=temp_path, quiet=True)
    df = pd.read_parquet(temp_path)
finally:
    if os.path.exists(temp_path):
        os.remove(temp_path)

display(df.head())
display(df.info())

,MÊS COMPETÊNCIA,MÊS REFERÊNCIA,UF,CÓDIGO MUNICÍPIO SIAFI,NOME MUNICÍPIO,CPF FAVORECIDO,NIS FAVORECIDO,NOME FAVORECIDO,VALOR PARCELA
0,202601,202601,BA,3641,ITIUBA,***.557.525-**,1.626318e+10,GERCINA SANTOS SOUZA,"600,00"
1,202601,202601,RJ,6001,RIO DE JANEIRO,***.986.507-**,1.648093e+10,LUCIANA DA SILVA FERREIRA,"650,00"
2,202601,202601,AL,2705,ARAPIRACA,None,1.616633e+10,DAYANE RODRIGUES DE SOUSA,"750,00"
3,202601,202601,PB,1959,BREJO DOS SANTOS,***.577.104-**,1.606359e+10,KESILY PEREIRA DE MEDEIROS,"750,00"
4,202601,202601,GO,9489,NIQUELANDIA,***.672.381-**,2.106645e+10,AMANDA KARITA DA SILVA,"800,00"


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7749667 entries, 0 to 7749666
Data columns (total 9 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   MÊS COMPETÊNCIA         int64  
 1   MÊS REFERÊNCIA          int64  
 2   UF                      object 
 3   CÓDIGO MUNICÍPIO SIAFI  int64  
 4   NOME MUNICÍPIO          object 
 5   CPF FAVORECIDO          object 
 6   NIS FAVORECIDO          float64
 7   NOME FAVORECIDO         object 
 8   VALOR PARCELA           object 
dtypes: float64(1), int64(3), object(5)
memory usage: 532.1+ MB


None

### Questão 1
Quantas linhas existem no `df`? Responda com **um número inteiro**.


In [17]:
df.shape

(7749667, 9)

- Existem 7749667 linhas

### Questão 2
Quantas colunas existem no `df`? Responda com **um número inteiro**.


- Existem 9 colunas

### Questão 3
Quantos registros têm `cpf_favorecido` preenchido (não vazio e diferente de `nan`)? Responda com **um número inteiro**.


In [18]:
display(len(df) - df.isnull().sum())

#retirando as linhas com valores nulos
df = df.dropna()

MÊS COMPETÊNCIA           7749667
MÊS REFERÊNCIA            7749667
UF                        7749667
CÓDIGO MUNICÍPIO SIAFI    7749667
NOME MUNICÍPIO            7749667
CPF FAVORECIDO            6101351
NIS FAVORECIDO            7749534
NOME FAVORECIDO           7749667
VALOR PARCELA             7749667
dtype: int64

- Há 6101351 cpf favorecido

### Questão 4
Quantos valores **distintos** existem em `cpf_favorecido` (considerando apenas registros preenchidos)? Responda com **um número inteiro**.


In [19]:
len(df["CPF FAVORECIDO"].unique())

960416

- Tem 960416 cpf favorecido distintos

### Questão 5
Quantos `nis_favorecido` **distintos** existem? Responda com **um número inteiro**.


In [20]:
len(df["NIS FAVORECIDO"].unique())

5916822

- Tem 7515189 nis favorecido distintos

### Questão 6
Quantos municípios **distintos** aparecem em `nome_municipio`? Responda com **um número inteiro**.


In [21]:
len(df["NOME MUNICÍPIO"].unique())

5291

- Tem 5291 municípios distintos

### Questão 7
Quantas UFs **distintas** aparecem em `uf`? Responda com **um número inteiro**.


In [22]:
len(df["UF"].unique())

27

- Tem 27 UFs distintas, ou seja, todas

### Questão 8
Qual é a **média** de `valor_parcela`? Responda com **um número decimal**.


In [23]:
#convertendo a coluna "VALOR PARCELA" para float
df["VALOR PARCELA"] = df["VALOR PARCELA"].str.replace(',','.').astype(float)

df["VALOR PARCELA"].mean()

667.2742254953042

- A media é aproximadamente R$667,27

### Questão 9
Qual é a **mediana** de `valor_parcela`? Responda com **um número decimal**.


In [24]:
df["VALOR PARCELA"].median()

650.0

- A mediana é R$650,00

### Questão 10
Qual é o **valor mínimo** de `valor_parcela`? Responda com **um número decimal**.


In [25]:
df["VALOR PARCELA"].min()

25.0

- O menor valor parcela é R$25,00

### Questão 11
Qual é o **valor máximo** de `valor_parcela`? Responda com **um número decimal**.


In [26]:
df["VALOR PARCELA"].max()

3514.0

- O maior valor parcela é R$3514,00

### Questão 12
Faça a **contagem de registros por UF** e informe a maior contagem encontrada. Responda com **um número inteiro**.


In [27]:
df["UF"].value_counts()

UF
BA    784522
SP    666902
PE    497378
CE    468945
MG    456960
RJ    417880
PA    410652
MA    387124
PB    212469
PI    188399
AM    187296
AL    170898
PR    169637
RS    162741
RN    162318
GO    138930
SE    121237
ES     88921
MT     68043
SC     58162
MS     57341
DF     48908
TO     47255
AC     38569
RO     36732
AP     34827
RR     18305
Name: count, dtype: int64

- Bahia é a UF com mais registros, 784522

### Questão 13
Calcule a **média de `valor_parcela` por UF** e informe a maior média encontrada. Responda com **um número decimal**.


In [28]:
df.groupby("UF")["VALOR PARCELA"].mean()

UF
AC    719.444372
AL    676.326358
AM    722.582538
AP    722.617797
BA    659.324125
CE    659.416580
DF    670.487487
ES    654.669145
GO    661.730641
MA    690.841314
MG    654.063734
MS    680.379920
MT    680.862646
PA    691.528279
PB    665.623136
PE    662.788189
PI    666.448214
PR    658.199656
RJ    660.256327
RN    658.477655
RO    667.543259
RR    736.430374
RS    657.386928
SC    657.769523
SE    662.728004
SP    652.938091
TO    682.572257
Name: VALOR PARCELA, dtype: float64

- Acre é a UF com maior média, R$719,44

### Questão 14
Calcule a **mediana de `valor_parcela` por município** e informe a maior mediana encontrada. Responda com **um número decimal**.


In [29]:
df.groupby("NOME MUNICÍPIO")["VALOR PARCELA"].median()

NOME MUNICÍPIO
ABADIA DE GOIAS        700.0
ABADIA DOS DOURADOS    650.0
ABADIANIA              650.0
ABAETE                 650.0
ABAETETUBA             650.0
                       ...  
XIQUE-XIQUE            650.0
ZABELE                 650.0
ZACARIAS               700.0
ZE DOCA                650.0
ZORTEA                 650.0
Name: VALOR PARCELA, Length: 5291, dtype: float64

- O município Abadida de Goias é o municipio com a maior mediana, R$700,00

### Questão 15
Para cada `mes_referencia`, calcule o **máximo** de `valor_parcela`. Informe o maior máximo encontrado. Responda com **um número decimal**.


In [30]:
df.groupby("MÊS REFERÊNCIA")["VALOR PARCELA"].max()

MÊS REFERÊNCIA
202502     750.0
202505     800.0
202506    2304.0
202507    2070.0
202508    2304.0
202509    2020.0
202510    2212.0
202511    2930.0
202512    2930.0
202601    3514.0
Name: VALOR PARCELA, dtype: float64

- O maior valor encontrado foi em 01/2026, sendo R$3514,00